[SR-12311] Improve diagnostic for read-only properties

HassanElDesouky opened the pull request:
<!-- What's in this pull request? -->
I changed the diagnostic message for protocol read-only properties as suggested in [[SR-12311]](https://bugs.swift.org/browse/SR-12311)

This is my first contribution to Swift so if you think that I understood the problem in a wrong way or did something wrong, please guide me towards the solution.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [[SR-12311]](https://bugs.swift.org/browse/SR-12311) <br>
Radar Url: rdar://problem/60013553

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

HassanElDesouky commented:
@varungandhi-apple Please review

HassanElDesouky commented:
@swift-ci please smoke test

xwu commented:
I suspect you'll need to update any tests for this to pass testing. After that, someone with permissions can trigger the smoke test.

HassanElDesouky commented:
> I suspect you'll need to update any tests for this to pass testing.

There were no tests regarding this diagnostic message. 

> After that, someone with permissions can trigger the smoke test.

Thank you, I didn't know that!

theblixguy commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1973,7 +1973,7 @@ WARNING(protocol_usable_from_inline_warn,none,
         "should be '@usableForInline' or public", (bool))
 ERROR(protocol_property_must_be_computed_var,none,
-      "immutable property requirement must be declared as 'var' with a "
+      "protocols cannot require properties to be immutable; declare read-only properties by using 'var' with a "
```
It doesn’t look like this is within 80 columns, could you please fix that?

HassanElDesouky replied:
Sorry, I didn't notice that. I fixed it now.

theblixguy commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1973,7 +1973,7 @@ WARNING(protocol_usable_from_inline_warn,none,
         "should be '@usableForInline' or public", (bool))
 ERROR(protocol_property_must_be_computed_var,none,
-      "immutable property requirement must be declared as 'var' with a "
+      "protocols cannot require properties to be immutable; declare read-only properties by using 'var' with a "
```
Could you please add some tests for this (or update the existing ones if you haven’t)?

HassanElDesouky replied:
@theblixguy I updated the test files. Can you check now if everything is okay?

theblixguy replied:
Looks great! Thank you!

xwu commented:
> > I suspect you'll need to update any tests for this to pass testing.
> 
> There were no tests regarding this diagnostic message.

There is definitely at least one. Hint: you can search this repository for the old diagnostic text.

LucianoPAlmeida commented:
@HassanElDesouky It may a good thing to squash those changes into a single commit :))

HassanElDesouky committed:
Improve diagnostic for read-only properties
Make the message within 80 columns width
Improve diagnostic for read-only properties
Improve diagnostic for read-only properties
Improve diagnostic for read-only properties

HassanElDesouky commented:
> @HassanElDesouky It may a good thing to squash those changes into a single commit :))

Thank you for pointing that out I just squashed the changes into a single commit.

harlanhaskins commented:
Welcome to the project, and thanks for your contribution! These changes look good to me. I’ll go ahead and get CI started.

@swift-ci please test.

HassanElDesouky commented:
> Welcome to the project, and thanks for your contribution! 

Thank you @harlanhaskins , I really couldn't have done it without the help of the community. @theblixguy's blog helped me a lot to start my first PR. 

> I’ll go ahead and get CI started.

The CI finished successfully, does this mean it's able to merge?

harlanhaskins commented:
Merged. Thanks again for your contribution!

CodaFi commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1973,8 +1973,8 @@ WARNING(protocol_usable_from_inline_warn,none,
         "protocol %select{refined|used}0 by '@usableFromInline' protocol "
         "should be '@usableForInline' or public", (bool))
 ERROR(protocol_property_must_be_computed_var,none,
```
The next thing to do is split the dependent clause off into a note and attach the fixit there. This diagnostic is quite wordy now...

varungandhi-apple commented:
Thanks for the contribution @HassanElDesouky! 😄
